### PR TITLE
Support nested fields and arrays

### PR DESCRIPTION
**Pull request by williamkhshea**
See #47 

In input form

```
{getFieldDecorator('member[0].name.firstname', {})(<input/>)}
{getFieldDecorator('member[1].name.firstname', {})(<input/>)}
```

this.props.form.getFieldsValue() 返回数据结构为
```
members: [{ name: {firstname: "xxx"} }, { name: {firstname: "yyy"} }]
```

### CODE REVIEW

**Comment by alexggordon** [comment]
@williamkhshea just saw this. Upon validate, does this also return the same nested object?

IE, on submit, will it return 

```
{member: [
     {name: {firstname: 'fdsafdsa'}},
     {name: {firstname: 'firstname_two'}}
]}
```

If it's doing this, then I'll close my PR in favor of this one.

**Comment by williamkhshea** [comment]
@alexggordon Yes, and errors will be nested object as well.

I have added an example called "nested-field" in the example folder. You can check it out and see what how the outputs look like.

**Comment by alexggordon** [comment]
Sweet job @williamkhshea.

**Comment by afc163** [comment]
:+1:

File: package.json [hunk 1]
```diff
@@ -62,6 +62,7 @@
     "babel-runtime": "6.x",
     "dom-scroll-into-view": "1.x",
     "hoist-non-react-statics": "1.x",
+    "lodash": "^4.17.2",
```
Comment (yiminghe) [inline package.json]: do not use lodash(especially too big for mobile), or at least use lodash-xx

**Review by yiminghe — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (williamkhshea) [inline package.json]: @yiminghe OK, I will update to use custom build of lodash

**Comment by williamkhshea** [comment]
@yiminghe How can I ignore the lint style checking for the lodash custom build file before I commit?

**Comment by yiminghe** [comment]
I mean something like https://www.npmjs.com/package/lodash.get

**Comment by williamkhshea** [comment]
@yiminghe Dependencies updated. Please review.

**Comment by afc163** [comment]
Use https://github.com/mariocasciaro/object-path ?

**Comment by williamkhshea** [comment]
@afc163 Object-path uses only dot notation for array path (i.e. 'a.0'), unlike lodash which also supports bracket notation (i.e. 'a[0]'). It creates problems when you want to have numbers as keys of an object, says in case of IDs. It gives you an array instead of object with keys as IDs.

If you want to trim down the size of the libraries further, I would suggest you do a custom build on lodash get,has,set function and make an npm package, and I can use it as dependencies.

File: tests/overview.spec.js [hunk 2]
```diff
@@ -137,4 +189,100 @@ describe('overview usage', () => {
...
+    Simulate.change(form.getFieldInstance('foo.a.x'));
+    expect(isEqual(form.getFieldValue('foo'), {
+      a: {
+        x: '9',
+        y: '6',
+      },
+      b: ['7', '8'],
+    })).to.be(true);
+    form.resetFields();
+    expect(isEqual(form.getFieldValue('foo'), {
```
Comment (yiminghe) [inline tests/overview.spec.js]: expect(form.getFieldValue('foo')).to.eql({....})

**Review by yiminghe — CHANGES_REQUESTED** [review]
last change. please squash into one commit. ready to merge


thank you

File: tests/overview.spec.js (hunk 2 above)
Comment (williamkhshea) [inline tests/overview.spec.js]: @yiminghe expect(...).to.eql does not handle sparse arrays (e.g. [undefined, 123]). See screenshot below.

2016-12-22 16_15_58-will s imac - teamviewer - free license non-commercial use only

Comment (yiminghe) [inline tests/overview.spec.js]: ok, squash into one commit, and force push please

Comment (williamkhshea) [inline tests/overview.spec.js]: @yiminghe done

**Comment by williamkhshea** [comment]
@yiminghe I should have cleaned up my code before submitting. Thanks for helping out.

**Comment by acmeid** [comment]
我在使用嵌套写法的时候，并没有生效：
```
<FormItem label="数量" colon={false}>
  {getFieldDecorator('services[0].num', {
      rules: [{ required: true, message: '请输入数量' }],
   })(
      <InputNumber />
   )}
</FormItem>

services: [{
    num: {value: 1}
}]
```

**Comment by mkrn** [comment]
@williamkhshea Can this bug be related to this PR? 
https://github.com/ant-design/ant-design/issues/5835
Can you please give me a hint if you have this situation working (array of objects)

**Comment by nanxiaobei** [comment]
then setFieldsValue not work with nested assign, anyone know how to solve?

**Comment by benjycui** [comment]
@nanxiaobei please file a new issue and provide re-producible demo http://codepen.io/benjycui/pen/NdQdBw?editors=0011

**Comment by warmhug** [comment]
`isFieldTouched` seems also not work with nested.

**Comment by zhangpeng11** [comment]
我注释掉getFieldValue这种代码以后神奇的就变回数组了，这是什么情况

**Comment by jarforjs** [comment]
> then setFieldsValue not work with nested assign, anyone know how to solve?

me too,then i use setFiledsValue(()=>{setTimeout...}),Nevertheless, it seems to be successful, but there are still problems

**Comment by 17601617393** [comment]
> 参见[＃47](https://github.com/react-component/form/issues/47)
> 
> 输入形式
> 
> ```
> {getFieldDecorator('member[0].name.firstname', {})(<input/>)}
> {getFieldDecorator('member[1].name.firstname', {})(<input/>)}
> ```
> 
> this.props.form.getFieldsValue（）返回数据结构为
> 
> ```
> members: [{ name: {firstname: "xxx"} }, { name: {firstname: "yyy"} }]
> ```
id再转一下string类型就OK了

**Comment by Jiang-Xuan** [comment]
如果有人想要 key 里面含有 . 的话, 参考 https://github.com/lodash/lodash/issues/3529#issuecomment-448586433

**Comment by pkuphy** [comment]
楼上的方法可以，效果是这样的：

`getFieldDecorator('["user.name"]', ...`

Screenshot-2023-05-19_16_43_22

**Pull request merged**